image-prune

```diff
@@ -152,34 +140,74 @@ def prune_images(force, dryrun, quiet=False, open_pull_requests=True, offline=Fa
             target = os.readlink(entry.path)
             targets.add(target)
 
-    expiry_threshold = now - IMAGE_EXPIRE * 86400
-    for entry in os.scandir(get_images_data_dir()):
-        if not force and (enough_disk_space() and entry.stat(follow_symlinks=False).st_mtime > expiry_threshold):
-            continue
-        if entry.is_file() and \
-                any(entry.name.endswith(ext) for ext in ['.qcow2', '.partial']) and \
-                entry.name not in targets:
-            if not quiet or dryrun:
-                sys.stderr.write("Pruning {0}\n".format(entry.name))
+    return targets
+
+
+class ImageCache:
+    def prune(self, keepers, force=False, dryrun=False, quiet=False):
+        expiry_threshold = time.time() - IMAGE_EXPIRE * 86400
+
+        for image, mtime in self.list_files():
+            if not any(image.endswith(ext) for ext in ['.qcow2', '.partial']):
+                debug(f'Skipping file {image} with unknown extension')
+                continue
+
+            if image in keepers:
+                debug(f'Skipping image {image} which is in the keepers list')
+                continue
+
+            if not force and self.enough_space() and mtime > expiry_threshold:
+                debug(f'Skipping image {image} which is new enough (and not forced or low on space)')
+                continue
+
+            if not quiet:
+                print(f"Pruning {image}", file=sys.stderr)
+
             if not dryrun:
-                os.unlink(entry.path)
+                self.delete_image(image)
+
+        self.cleanup(quiet, dryrun)
+
+
+class LocalImageDirectory(ImageCache):
+    def __init__(self, directory=None):
+        self.directory = directory or get_images_data_dir()
+        self.space_threshold = float(os.environ.get("PRUNE_THRESHOLD_G", 15))
+
+    def list_files(self):
+        for entry in os.scandir(self.directory):
+            if entry.is_file(follow_symlinks=False):
+                yield entry.name, entry.stat().st_mtime
+
+    def delete_file(self, filename):
+        os.unlink(os.path.join(self.directory, filename))
+
+    def enough_space(self):
+        st = os.statvfs(self.directory)
+        free = st.f_bavail * st.f_frsize / (1024 * 1024 * 1024)
+        return free >= self.space_threshold
+
+    def cleanup(self, quiet, dryrun):
+        # clean any broken symlinks in IMAGES_DIR
```

Comment:
I'm not sure why we need that code still -- the image cache dir *never* gets symlinks. The links are maintained in the bots.git. The task containers clean up their bots checkout for each iteration anyway, and if you want to clean up your local checkout, then `git clean -fdx` is a much better way IMHO. So from my POV, feel free to toss this.

Reply:
This is indeed about the local working directory, but it's kinda incomplete: for example, it won't clean up other worktrees, or checkouts in other places.  Even as I was doing this I was thinking "this needs a separate tool...".

I'll drop it for now.